### ISSUE TITLE
Best way to check why a function is recompiled under `filter_jit`?

### ISSUE THREAD

**Issue by LouisDesdoigts** [patrick-kidger/equinox#640]
So I have a step function in the same vein as the [mnist example](https://docs.kidger.site/equinox/examples/mnist/) using a `filter_jit` decorator. However this step function is compiling _twice_. I assume this is due to some type promotion on a leaf, but I can not seem to figure out _which_ one is causing this issue. 

If I understand correctly, `filter_jit` partitions the input pytree into 'static' and 'dynamic' using the `is_array` filter function. The leaves of the static pytree have a hash that is applied to `jit` under the hood. Given this, I have tried three checks to see what is causing the recompile, none of which show differences. 

The basic step function looks like this:

```python
@eqx.filter_jit
def step_fn(model, opt_state, args):
    loss, grads = loss_fn(model, args)
    updates, opt_state = optim.update(grads, opt_state, model)
    model = eqx.apply_updates(model, updates)
    return model, loss, opt_state
```

I have tried to check for differences in the pytree a few different ways. First the shape and dtype of the leaves of the dynamic pytree.

```python
fspec = jax.tree_map(eqx.is_array, model)

# Get the 'before' and 'after' pytree
before = eqx.filter(model, fspec)
model, loss, opt_state = step_fn(model, opt_state, args)
after = eqx.filter(model, fspec)

# Check fucntion
def check_fn(l1, l2):
    if (l1.shape != l2.shape) or (l1.dtype != l2.dtype):
        print("DIFFERENT")
_ = jax.tree_map(check_fn, before, after) # Prints nothing!

# Always recompiles a second time
model, loss, opt_state = step_fn(model, opt_state, args)
```

I have also tried checking the hash of each leaf of the static pytree, along with a check using `tree_equal`.

```python
fspec = jax.tree_map(lambda leaf: not eqx.is_array(leaf), model)

# Get the 'before' and 'after' pytree
before = eqx.filter(model, fspec)
new_model, loss, opt_state = step_fn(model, opt_state, args)
after = eqx.filter(new_model, fspec)

# Check fucntion
def check_fn(l1, l2):
    if hash(l1) != hash(l2):
        print("DIFFERENT")
_ = jax.tree_map(check_fn, before, after) # Prints nothing!
print(eqx.tree_equal(before, after)) # Prints 'True'

# Always recompiles a second time
model, loss, opt_state = step_fn(model, opt_state, args)
```

---

So given that these don't highlight the issue, I was wondering what (likely obvious) thing I have missed, or what I am miss-understanding about what triggers a re-compile!

Thanks for the help!

**Comment by patrick-kidger** [comment]
Can you try `eqx.debug.assert_max_traces(max_traces=1)`?

This will attempt to detect what changed that caused recompilation.

**Comment by LouisDesdoigts** [comment]
Wow as per-usual you already have a function to do exactly what I'm looking for, this is awesome! I'll write a little summary here in case anyone else has a similar problem in the future. 

Turns out it was not the model itself, but the optax `opt_state` object. My models typically need either different optimisers or no optimiser on specific leaves. Initially I was using calling `opt_state = optim.init(eqx.filter(model, eqx.is_array))` to initialise the `opt_state` object, however since some leaves do not have gradients taken with respect to them, their leaves are `None`. The `opt_state` object then has a different pytree structure when returned from the `step_fn`, causing the recompile.

The fix was to call `optim.init(...)` on a pytree that has `None` on all leaves with no gradients taken, ie matching the leaf structure of the `grad` pytree that is passed to `optim.update` inside the `step_fn`.

Thanks again for the help!
